### ISSUE TITLE
[stable/redis] Sentinels fails after cluster disaster

### ISSUE THREAD

**Issue by yusufgungor** [helm/charts#19393]
Hi,

We are using a test k8s cluster on AWS and shutting down the all nodes (workers, masters) when not using them.

Shutting down the cluster (all workers and masters) while redis with sentinel are running. After starting up the cluster then all sentinels are hanged with the similar log below:

```
1:X 05 Dec 2019 07:28:14.888 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:X 05 Dec 2019 07:28:14.888 # Redis version=5.0.6, bits=64, commit=00000000, modified=0, pid=1, just started
1:X 05 Dec 2019 07:28:14.888 # Configuration loaded
1:X 05 Dec 2019 07:28:14.892 * Running mode=sentinel, port=26379.
1:X 05 Dec 2019 07:28:14.892 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
1:X 05 Dec 2019 07:28:14.892 # Sentinel ID is f2cc9f1484773e0b43f68cda44d5d7c8dbe161ad
1:X 05 Dec 2019 07:28:14.892 # +monitor master test-redis-sentinel-master-set 100.114.31.245 6379 quorum 2
1:X 05 Dec 2019 07:29:14.972 # +sdown master test-redis-sentinel-master-set 100.114.31.245 6379
1:X 05 Dec 2019 07:29:14.972 # +sdown slave 100.106.63.159:6379 100.106.63.159 6379 @ test-redis-sentinel-master-set 100.114.31.245 6379
1:X 05 Dec 2019 07:29:14.972 # +sdown slave 100.115.204.234:6379 100.115.204.234 6379 @ test-redis-sentinel-master-set 100.114.31.245 6379
1:X 05 Dec 2019 07:29:14.972 # +sdown slave 100.114.31.248:6379 100.114.31.248 6379 @ test-redis-sentinel-master-set 100.114.31.245 6379
1:X 05 Dec 2019 07:29:14.972 # +sdown sentinel 3ad986f6f134526435c84e8507c30e0d11cc3566 100.114.31.245 26379 @ test-redis-sentinel-master-set 100.114.31.245 6379
1:X 05 Dec 2019 07:29:14.972 # +sdown sentinel ce92e0db09ba79602b3091e6a3a742da24915a5d 100.115.204.234 26379 @ test-redis-sentinel-master-set 100.114.31.245 6379
1:X 05 Dec 2019 07:29:14.972 # +sdown sentinel db02b9f1c760972c399d81903c39bcbc347b1467 100.106.63.159 26379 @ test-redis-sentinel-master-set 100.114.31.245 6379
```

These ip addresses are the old ip addresses for pods. (Before shutting down the cluster),

All sentinels print the similar logs with old ip addresses. All sentinels accept the all other components are down and do nothing.

After starting up a new cluster then only the new master is running but none of the sentinels detect it. Slaves also tries to access master with old ip address and fails. 

If we rollout restart the slaves and master statefulsets then it starts working.

Is that expected behaviour? Is it possible to let redis work as before shutting down the cluster?

Thanks.

**Comment by alemorcuq** [comment]
Hi, @yusufgungor.

This PR #19059 was merged recently addressing what I believe is the same issue your are having. Can you try enabling the `sentinel.staticID` parameter and let me know if that fixes it?

Regards,
Alejandro

**Comment by yusufgungor** [comment]
Hi @alemorcuq thanks for reply.

I am going to test it today, i'll inform you after tests.

**Comment by alemorcuq** [comment]
Is there any update on this, @yusufgungor ?

**Comment by peter-c-larsson** [comment]
Had the same issue and tested the static id. It seems when a pod is restarted and gets a new IP the pod that is restarted gets another sentinel entry going from 2 -> 3. (In the case with 1 master and 2 slaves)
This causes the same problem even with static id.

What I tested was to increase the number of slaves to 4 and then it could still select a new master even with that extra entry. So static id helps and doesn't solve it completely.

There is another issue as well as I noticed if the a node crashes and once the node starts. The pod gets restarted on the same node it keep the filesystem (with old ip in filesystem) and receive a new IP so there is a mismatch in the config and the ip requiring the existing pod to be deleted in order for the pod  to recover.

**Comment by peter-c-larsson** [comment]
And some logs from our test cluster with static ID, 1 master and 4 slaves and some restart of the pods. Notice how the master has one extra entry with it's old IP. That was the first restart and it hasn't happened for the other slaves after restarting them.
```
root@kube-master-1:~# kubectl exec -it -c sentinel staging-gitlab-redis-slave-1 -- redis-cli  -p 26379 SENTINEL sentinels gitlabmaster | grep ip -A 1 -B 3
1)  1) "name"
    2) "d6555f42b6c5d99b3fd1dfce36d613fda05ec534"
    3) "ip"
    4) "10.20.209.190"
--
     28) "0"
2)  1) "name"
    2) "551cd3cc716f644155d4ead79333803459399a1d"
    3) "ip"
    4) "10.20.201.113"
--
     28) "0"
3)  1) "name"
    2) "632af3021de4d4e6267a1deee5cf3e07556f9cf4"
    3) "ip"
    4) "10.20.245.190"
--
     28) "0"
4)  1) "name"
    2) "7ab0c98f42696afdea402c6a5a488098229513a2"
    3) "ip"
    4) "10.20.229.41"
root@kube-master-1:~#
root@kube-master-1:~#
root@kube-master-1:~# kubectl exec -it -c sentinel staging-gitlab-redis-master-0 -- redis-cli  -p 26379 SENTINEL sentinels gitlabmaster | grep ip -A 1 -B 3
1)  1) "name"
    2) "632af3021de4d4e6267a1deee5cf3e07556f9cf4"
    3) "ip"
    4) "10.20.245.190"
--
     28) "0"
2)  1) "name"
    2) "551cd3cc716f644155d4ead79333803459399a1d"
    3) "ip"
    4) "10.20.201.113"
--
     28) "0"
3)  1) "name"
    2) "d6555f42b6c5d99b3fd1dfce36d613fda05ec534"
    3) "ip"
    4) "10.20.209.190"
--
     28) "0"
4)  1) "name"
    2) "7ab0c98f42696afdea402c6a5a488098229513a2"
    3) "ip"
    4) "10.20.229.127"
--
     30) "0"
5)  1) "name"
    2) "53f756f0165e9925b626c423eb3c5f86c2e391bf"
    3) "ip"
    4) "10.20.158.235"
root@kube-master-1:~# kubectl exec -it -c sentinel staging-gitlab-redis-slave-2 -- redis-cli  -p 26379 info | grep master0 && kubectl get pods -lapp=redis -o wide
master0:name=gitlabmaster,status=ok,address=10.20.209.190:6379,slaves=5,sentinels=5
NAME                            READY   STATUS    RESTARTS   AGE     IP              NODE          NOMINATED NODE   READINESS GATES
staging-gitlab-redis-master-0   2/2     Running   0          4m12s   10.20.229.41    kube-node-5   <none>           <none>
staging-gitlab-redis-slave-0    2/2     Running   4          22h     10.20.245.190   kube-node-2   <none>           <none>
staging-gitlab-redis-slave-1    2/2     Running   0          3m11s   10.20.158.235   kube-node-1   <none>           <none>
staging-gitlab-redis-slave-2    2/2     Running   0          13m     10.20.201.113   kube-node-3   <none>           <none>
staging-gitlab-redis-slave-3    2/2     Running   0          22h     10.20.209.190   kube-node-4   <none>           <none>
```

**Comment by javsalgar** [comment]
Hi,

Would it make sense that, before starting sentinel, check if there's an entry with the same static id and have it removed? I'm not sure if that would fix the duplicity in the masters.
